### PR TITLE
[global-hooks] Fix cluster dns address discovery

### PR DESCRIPTION
**Pull request by EvgeniiAl**
Take cluster dns address from service with `kube-dns` name if there are multiply clusterIp services labeled `k8s-app: kube-dns` in cluster.

E.g.:  enable node-local-dns plugin in GKE, to deploy two services with same labels `k8s-app: kube-dns` and selectors  named `kube-dns` and `kube-dns-upstream`,
after that cluster dns address fails with error `"ERROR: can't find a single kube-dns service`

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: global-hooks
type: fix
summary:  Fix cluster dns address discovery
impact_level: default
```

### CODE REVIEW

File: global-hooks/discovery/cluster_dns_address.go [hunk 1]
```diff
@@ -25,10 +25,30 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
+const (
+	dnsAddressServiceSnapName = "kube_dns_cluster_ip"
```
Comment (name212) [inline global-hooks/discovery/cluster_dns_address.go:29]: Should removed

File: global-hooks/discovery/cluster_dns_address.go [hunk 2]
```diff
@@ -25,10 +25,30 @@ import (
...
+	dnsAddressByD8AppSnapName = "cluster_dns_address_by_d8s_app"
+)
+
 var _ = sdk.RegisterFunc(&go_hook.HookConfig{
 	Kubernetes: []go_hook.KubernetesConfig{
 		{
-			Name:       "dns_cluster_ip",
+			Name:       dnsAddressServiceSnapName,
```
Comment (name212) [inline global-hooks/discovery/cluster_dns_address.go:36]: We do not need this binding

File: global-hooks/discovery/cluster_dns_address.go [hunk 3]
```diff
@@ -50,31 +70,58 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
...
-		addr := addrRaw.(string)
-		if addr == "None" || addr == "" {
+
+	for _, sRaw := range input.Snapshots["dns_cluster_ip"] {
+		s := sRaw.(ServiceAddr)
+
+		if s.Name == "kube-dns" && s.ClusterIP != "None" && s.ClusterIP != "" {
+			input.Values.Set("global.discovery.dnsAddress", s.ClusterIP)
+			return nil
+		}
```
Comment (name212) [inline global-hooks/discovery/cluster_dns_address.go:114]: Let's remove it and move logic below

File: global-hooks/discovery/cluster_dns_address.go [hunk 4]
```diff
@@ -50,31 +70,58 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
...
+
+		if s.ClusterIP == "None" || s.ClusterIP == "" {
 			continue
 		}
```
Comment (name212) [inline global-hooks/discovery/cluster_dns_address.go:119]:
```
if s.Name == "kube-dns" {
  dnsAddress = s.ClusterIP
  break
}
```

Comment (name212) [inline global-hooks/discovery/cluster_dns_address.go:119]:
```suggestion
if s.Name == "kube-dns" {
  dnsAddress = s.ClusterIP
  break
}
```

Comment (evgenii-aladin) [reply global-hooks/discovery/cluster_dns_address.go:93]: done

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/openstack'...

**Comment by name212** [comment]
/e2e/abort openstack;Standard;containerd;1.23 4870684687 4870684687-1-con-1-23 /sys/deckhouse-oss/install:pr4521 debian@45.12.229.167

**Comment by name212** [comment]
/e2e/abort openstack;Standard;containerd;1.23 4870684687 4870684687-1-con-1-23 /sys/deckhouse-oss/install:pr4521 debian@45.12.229.167

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've recognized your '/e2e/abort' command and started the workflow...

**Review by name212 — APPROVED** [review]

**Review by nabokihms — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @nabokihms. I've started the workflow for label 'e2e/run/eks'...

**Comment by nabokihms** [comment]
/e2e/abort eks;WithoutNAT;containerd;1.23 4960206717 4960206717-1-con-1-23 /sys/deckhouse-oss/install:pr4521

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @nabokihms. I've recognized your '/e2e/abort' command and started the workflow...

**Pull request merged**